enhancements/ingress/dynamic-config-manager.md

```diff
@@ -0,0 +1,469 @@
+long-lived connections._
+
+Without the Dynamic Config Manager, OpenShift router suffers from a well known
+performance issue when the following two conditions are met:
+
+* A router pod reloads its configuration frequently because of route or endpoints updates.
+* The same router pod handles long-lived connections.
+
+Reloading the HAProxy configuration involves forking a new process with the new
+configuration.  The old process remains open until all connections that were
+open at the time of the configuration reload have terminated.  In the case of
+long-lived connections, this means that the old process can remain for a long
+period of time.  If the router has frequent configuration reloads due to changes
+to route configuration, this can cause these old processes to accumulate and use
+a large amount of memory, on the order of hundreds of megabytes or multiple
+gigabytes per process, ultimately causing out-of-memory errors on the node host.
+
+DCM addresses this issue by reducing the need to reload the configuration.
+Instead, DCM configures HAProxy using its Unix domain control socket, which does
+not require forking a new HAProxy process.
+
+The degree to which DCM mitigates this issue is dependent on the nature of the
+configuration changes: Some changes still require a configuration reload, but
+the majority of changes can be performed through HAProxy's control socket.
+
+Initially, DCM will allow scale-out and scale-in of 1 server (pod endpoint) per
+backend (route).  Scaling out more than 1 server will still require a fork and
+reload initially.  In future iterations of the feature, DCM can be enhanced to
+enable scale-out and scale-in of arbitrarily many servers as pods are created
+and deleted, and also scale-out and scale-in of backends as routes are created
+and deleted, as well as changes to certificates and other route options, all
+without requiring a fork and reload.
+
+_As a cluster administrator, I want metrics to be accurate when routes and
+endpoints are updated._
+
+OpenShift router does take care to preserve metrics values across reloads in
+order to avoid resetting counters to zero when the new process starts.  However,
+while values from the old process are preserved, the old process cannot update
+metrics after the reload.  For example, the metrics reflect the total number of
+bytes that the old process had transferred at the point of the configuration
+reload, but if the old process continues transferring data after the new process
+starts, the metrics do not reflect the count of any additional bytes
+transferred.
+
+Dynamic Config Manager addresses this issue again by reducing the need to reload
+the configuration and fork a new process.  Again, the degree to which DCM
+mitigates this issue is dependent on the nature of the configuration changes.
+
+_As a project administrator, I want HAProxy to balance traffic evenly over old
+and new pods when I scale my application up._
```

Comment:
Yeah, I would say it's still true.  I prefer "evenly" because it is more evocative of the problem we are trying to avoid: an excessive share of the traffic going to the first server in the backend.  However, I can write "correctly" if that is clearer.  Adding "or down" does make sense.  (I should probably make it "in or out" since we are talking about adding or removing pod replicas, not increasing or decreasing the resources available to each pod.)